torch/_inductor/fx_passes/bucketing.py

```diff
@@ -277,523 +225,364 @@ def _rank_idx_dict(group_name: str) -> dict[int, int]:
+    )
+    wait_tensor = torch.ops.c10d_functional.wait_tensor(
+        torch.ops._c10d_functional.all_gather_into_tensor_out.default(
+            new_ag_in, group_size, group_name, out=new_ag_out
+        )
+    )
+    new_ag_out_reshaped = wait_tensor.reshape(group_size, -1)
+    outs = torch.split_with_sizes(
+        new_ag_out_reshaped,
+        ins_split_sizes,
+        dim=1,
+    )
+    outs_reshaped = [
+        o.reshape((shape[0] * group_size,) + shape[1:])
+        for o, shape in zip(outs, ins_sizes)
+    ]
+    return outs_reshaped
+
+
+def all_gather_merge_fn_to_trace_functional(
+    ag_ins: list[torch.Tensor],
+    group_size: int,
+    group_name: str,
+    dtype: torch.dtype,  # type: ignore[name-defined]
+    local_rank: int,
+) -> list[torch.Tensor]:
+    # Suboptimal functional implementation without mutations.
+    ins_sizes = [ag_in.shape for ag_in in ag_ins]
+    ins_split_sizes = [ag_in.numel() for ag_in in ag_ins]
+    ag_input_numel = sum(ins_split_sizes)
+    device = ag_ins[0].device
+    new_ag_out = torch.empty(ag_input_numel * group_size, dtype=dtype, device=device)
+    ag_ins_flattened = [ag_in.reshape(-1) for ag_in in ag_ins]
+    new_ag_in = torch.cat(ag_ins_flattened, dim=0)
+    new_ag_in = new_ag_out.narrow(0, ag_input_numel * local_rank, ag_input_numel)
+    foreach_copy_dsts = torch.split(new_ag_in, ins_split_sizes)
+    torch._foreach_copy_(foreach_copy_dsts, ag_ins_flattened)
+    wait_tensor = torch.ops.c10d_functional.wait_tensor(
+        torch.ops._c10d_functional.all_gather_into_tensor_out.default(
+            new_ag_in, group_size, group_name, out=new_ag_out
+        )
+    )
+    new_ag_out_reshaped = wait_tensor.reshape(group_size, -1)
+    outs = torch.split_with_sizes(
+        new_ag_out_reshaped,
+        ins_split_sizes,
+        dim=1,
+    )
+    outs_reshaped = [
+        o.reshape((shape[0] * group_size,) + shape[1:])
+        for o, shape in zip(outs, ins_sizes)
+    ]
+    return outs_reshaped
+
+
+def _trace(fn, inps) -> torch.fx.GraphModule:  # type: ignore[no-untyped-def]
+    fake_mode = maybe_get_fake_mode(inps[0][0])
+    assert fake_mode is not None
+    with fake_mode, enable_python_dispatcher():
+        return make_fx(fn)(*inps)
+
+
+def _insert_fn_trace_before_node(  # type: ignore[no-untyped-def]
+    g: torch.fx.Graph,
+    fn_to_trace,
+    inps,
+    insert_before_node,
+    g_fn_inps,
+    g_fn_outs,
+) -> dict[torch.fx.Node, torch.fx.Node]:  # type: ignore[no-untyped-def]
+    fn_gm = _trace(
+        fn_to_trace,
+        inps,
+    )
+    fn_g = fn_gm.graph
+    fn_g_ins = fn_g.find_nodes(op="placeholder")
+    env = {fn_g_ins[idx]: g_fn_inps[idx] for idx in range(len(g_fn_inps))}
+    g_fn_new_outs: list[torch.fx.Node] = []
+    with g.inserting_before(insert_before_node):
+        for _n in fn_g.nodes:
+            if _n.op == "placeholder":
+                continue
+            _new_n = g.node_copy(_n, lambda x: env[x])
```

Comment:
i'm kinda confused about this:
1) we do g.node_copy, which iiuc is putting the new node into the orig graph, then we erase it again
2) in the parent function (e.g. merge_all_gather) we again add the 'replacements' back into g

is this becuase in (1) we just want to use g.node_copy in a temporary manner to more easily build our 'env' dict and our 'g_fn_new_outs'?